crypto/http/http_client.c

```diff
@@ -840,12 +846,19 @@ BIO *OSSL_HTTP_transfer(const char *server, const char *port, const char *path,
 
     if (bio != NULL)
         cbio = bio;
+#ifndef OPENSSL_NO_SOCK
     else if ((cbio = HTTP_new_bio(server, port, proxy, proxy_port)) == NULL)
         return NULL;
+#else
+    else
+        return NULL;
+#endif
 
     (void)ERR_set_mark(); /* prepare removing any spurious libssl errors */
-    if (rbio == NULL && BIO_connect_retry(cbio, timeout) <= 0)
+#ifndef OPENSSL_NO_SOCK
+    if (bio == NULL && rbio == NULL && BIO_connect_retry(cbio, timeout) <= 0)
```

Comment:
Why did you add the `bio == NULL &&` part here? This leads to unintended behavior.
If `bio != NULL` and `rbio == NULL` this means, as documented, that the provided `bio` shall be used for connecting, writing request data, and reading response data.